x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -105,49 +111,125 @@ public void clusterChanged(ClusterChangedEvent event) {
         if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
             return;
         }
-        if (event.localNodeMaster() && shouldAllocateModels(event)) {
-            submitUnbatchedTask("allocating models to nodes", new ClusterStateUpdateTask() {
+        if (event.localNodeMaster() == false) {
+            return;
+        }
+
+        if (event.state().nodes().getMinNodeVersion().before(DISTRIBUTED_MODEL_ALLOCATION_VERSION)) {
+            // we should not try to rebalance assignments while there may be nodes running on a version
+            // prior to introducing distributed model allocation.
+            // But we should remove routing to removed or shutting down nodes.
+            removeRoutingToRemovedOrShuttingDownNodes(event);
```

Comment:
I know I am being paranoid here. But I want to be 100% sure. 

When a node is removed, routes are only rebalanced IFF one of the model's assigned to that node are no longer satisfied. 

I think this is always the case as we never over-allocate (so any loss is now no longer satisfied). Just wanted to capture my concern here :)

Reply:
Correct. The trick is in `TrainedModelAssignmentClusterService#shouldRebalanceModels`. We only return `true` if the removed node was hosting allocations. Indeed, that should be enough for us to rebalance as it would mean that at least that model is not satisfied as we do not over-allocate.